File: ios/MoltMarketplace/Core/Navigation/DeepLinkParser.swift
import Foundation

// MARK: - DeepLinkParser

/// Parses deep link URLs into Route values.
/// Supports `molt://` custom scheme and `https://molt.mt/` universal links.
enum DeepLinkParser {
    // MARK: - Internal

    /// Parses a URL into a Route. Returns nil for invalid or unrecognized links.
    static func parse(_ url: URL) -> Route? {
        let scheme = url.scheme?.lowercased() ?? ""
        let host = url.host?.lowercased() ?? ""

        switch scheme {
        case "molt":
            return parseMoltScheme(host: host, pathComponents: pathComponents(from: url))

        case "https" where host == "molt.mt":
            return parseUniversalLink(pathComponents: pathComponents(from: url))

        default:
            return nil
        }
    }

    // MARK: - Private

    private static func pathComponents(from url: URL) -> [String] {
        url.pathComponents.filter { $0 != "/" }
    }

    private static func parseMoltScheme(host: String, pathComponents: [String]) -> Route? {
        switch host {
        case "product":
            return parseProductRoute(pathComponents: pathComponents)

        case "category":
            return parseCategoryRoute(pathComponents: pathComponents)

        case "cart":
            return .cart

        case "order":
            return parseOrderRoute(pathComponents: pathComponents)

        case "profile":
            return .profile

        default:
            return nil
        }
    }

    private static func parseUniversalLink(pathComponents: [String]) -> Route? {
        guard let first = pathComponents.first else { return nil }

        switch first {
        case "product":
            return parseProductRoute(pathComponents: Array(pathComponents.dropFirst()))

        case "category":
            return parseCategoryRoute(pathComponents: Array(pathComponents.dropFirst()))

        default:
            return nil
        }
    }

    private static func parseProductRoute(pathComponents: [String]) -> Route? {
        guard let productId = pathComponents.first, !productId.isEmpty else { return nil }
        return .productDetail(productId: productId)
    }

    private static func parseCategoryRoute(pathComponents: [String]) -> Route? {
        guard let categoryId = pathComponents.first, !categoryId.isEmpty else { return nil }
        return .categoryProducts(categoryId: categoryId, categoryName: "")
    }

    private static func parseOrderRoute(pathComponents: [String]) -> Route? {
        guard let orderId = pathComponents.first, !orderId.isEmpty else { return nil }
        return .orderDetail(orderId: orderId)
    }
}

File: android/app/src/main/java/com/xirigo/ecommerce/core/navigation/DeepLinkParser.kt
package com.xirigo.ecommerce.core.navigation

import android.net.Uri

/**
 * Parses deep link URIs (xirigo:// and https://xirigo.com/) into Route objects.
 * Returns null for invalid or unrecognized links.
 */
object DeepLinkParser {

    fun parse(uri: Uri): Route? {
        val segments = extractPathSegments(uri) ?: return null
        return resolveRoute(segments)
    }

    private fun extractPathSegments(uri: Uri): List<String>? {
        val scheme = uri.scheme ?: return null
        return when (scheme) {
            "xirigo" -> uri.host?.let { host -> listOf(host) + uri.pathSegments }
            "https", "http" -> uri.host?.takeIf { it == "xirigo.com" }?.let { uri.pathSegments }
            else -> null
        }
    }

    private fun resolveRoute(segments: List<String>): Route? {
        if (segments.isEmpty()) return null
        val id = segments.getOrNull(1)?.takeIf { it.isNotBlank() }
        return when (segments.first()) {
            "product" -> id?.let { Route.ProductDetail(productId = it) }
            "category" -> id?.let { Route.CategoryProducts(categoryId = it, categoryName = "") }
            "cart" -> Route.Cart
            "order" -> id?.let { Route.OrderDetail(orderId = it) }
            "profile" -> Route.Profile
            else -> null
        }
    }
}
